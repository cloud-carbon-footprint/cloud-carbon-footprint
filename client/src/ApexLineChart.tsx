import React, { FunctionComponent } from 'react'
import Chart from 'react-apexcharts'
import moment from 'moment'

import { transformData } from './transformData'
import { EstimationResult } from './types'

export const ApexLineChart: FunctionComponent<ApexLineChartProps> = ({ data }) => {
  const timeSeriesData = transformData(data)
  const options = {
    series: [
      {
        name: 'co2e',
        data: timeSeriesData,
      },
    ],
    chart: {
      id: 'apexchart-example',
      toolbar: {
        tools: {
          zoomin: false,
          zoomout: false,
          customIcons: [],
        },
      },
    },
    height: '500px',
    markers: {
      size: 5,
    },
    stroke: {
      width: 2,
    },
    title: {
      text: 'Cloud Emissions',
      offsetY: -8,
      style: {
        fontSize: '24px',
      },
    },
    xaxis: {
      type: 'datetime',
      title: {
        text: '',
        offsetY: 8,
        style: {
          fontSize: '15px',
        },
      },
      labels: {
        formatter: function (value: any, timestamp: any, index: any) {
          return moment(new Date(timestamp)).format('DD-MMM-YY')
        },
      },
    },
    yaxis: {
      title: {
        text: 'CO2e (kg)',
        offsetX: -8,
        style: {
          fontSize: '15px',
        },
      },
    },
  }
  return <Chart options={options} series={options.series} type="line" height={options.height} />
}

type ApexLineChartProps = {
  data: EstimationResult[]
}
