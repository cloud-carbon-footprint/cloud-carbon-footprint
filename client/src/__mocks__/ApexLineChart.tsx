import React from 'react'

const ApexLineChart = () => <div>fake chart</div>

export { ApexLineChart }
