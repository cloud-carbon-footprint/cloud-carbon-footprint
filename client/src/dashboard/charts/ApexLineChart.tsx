import React, { FunctionComponent } from 'react'
import { renderToStaticMarkup } from 'react-dom/server'
import { useTheme } from '@material-ui/core/styles'
import { GetApp, PanTool, RotateLeft, ZoomIn } from '@material-ui/icons'
import Chart from 'react-apexcharts'
import moment from 'moment'

import { getChartColors } from '../../themes'
import { dailyTotals } from '../transformData'
import { EstimationResult } from '../../types'

export const ApexLineChart: FunctionComponent<ApexLineChartProps> = ({ data }) => {
  const theme = useTheme()

  // We need to get the HTML string version of these icons since ApexCharts doesn't take in custom React components.
  // Why, you might ask? Don't ask me, ask ApexCharts.
  const GetAppIconHTML = renderToStaticMarkup(<GetApp />)
  const PanToolIconHTML = renderToStaticMarkup(<PanTool />)
  const RotateLeftIconHTML = renderToStaticMarkup(<RotateLeft />)
  const ZoomInIconHTML = renderToStaticMarkup(<ZoomIn />)

  const co2SeriesData = dailyTotals(data).co2e
  const wattHoursSeriesData = dailyTotals(data).wattHours
  const CostSeriesData = dailyTotals(data).cost

  const colors = getChartColors(theme)
  const [blue, yellow, green] = [colors[0], colors[5], colors[8]]

  const options = {
    chart: {
      background: theme.palette.background.default,
      toolbar: {
        tools: {
          download: GetAppIconHTML,
          zoom: ZoomInIconHTML,
          zoomin: false,
          zoomout: false,
          pan: PanToolIconHTML,
          reset: RotateLeftIconHTML,
        },
      },
    },
    colors: [blue, yellow, green],
    height: '500px',
    markers: {
      size: 5,
    },
    series: [
      {
        name: 'AWS CO2e',
        data: co2SeriesData,
      },
      {
        name: 'AWS Watt Hours',
        data: wattHoursSeriesData,
      },
      {
        name: 'AWS Cost',
        data: CostSeriesData,
      },
    ],
    stroke: {
      width: 2,
    },
    theme: {
      mode: theme.palette.type,
    },
    title: {
      text: 'Cloud Usage',
      offsetY: -8,
      style: {
        fontSize: '24px',
      },
    },
    xaxis: {
      type: 'datetime',
      title: {
        text: '',
        offsetY: 8,
        style: {
          fontSize: '15px',
        },
      },
      labels: {
        formatter: function (value: any, timestamp: any, index: any) {
          return moment.utc(timestamp).format('DD-MMM-YY')
        },
      },
    },
    yaxis: [
      {
        title: {
          text: 'CO2e (kg)',
          offsetX: -8,
          style: {
            fontSize: '15px',
          },
        },
        decimalsInFloat: 3,
      },
      {
        title: {
          text: 'Watt hours (Wh)',
          opposite: -8,
          style: {
            fontSize: '15px',
            color: yellow,
          },
        },
        decimalsInFloat: 2,
        opposite: true,
        axisBorder: {
          show: true,
          color: yellow,
        },
        axisTicks: {
          show: true,
        },
      },
      {
        title: {
          text: 'Cost ($)',
          offsetX: -8,
          style: {
            fontSize: '15px',
            color: green,
          },
        },
        decimalsInFloat: 2,
        opposite: true,
        axisBorder: {
          show: true,
          color: green,
        },
        axisTicks: {
          show: true,
        },
      },
    ],
  }
  return <Chart options={options} series={options.series} type="line" height={options.height} />
}

type ApexLineChartProps = {
  data: EstimationResult[]
}
