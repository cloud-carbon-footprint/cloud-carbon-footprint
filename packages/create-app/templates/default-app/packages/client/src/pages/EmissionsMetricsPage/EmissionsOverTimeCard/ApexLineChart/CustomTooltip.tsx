/*
 * Â© 2021 Thoughtworks, Inc.
 */

/*
We're using a custom tool tip so that we can show data from the wattHours and cost series in the tooltip,
even when the wattHours and cost series are toggled off. When using the default tooltip provided by apex charts,
there is no way to access additional data in the series, for example, the wattHours and cost data in the co2e series.
The custom tooltip also allows us to add an asterisk for data points that were estimated using an average CPU constant.
 */

import React, { ReactElement } from 'react'
import moment from 'moment'
import { cloudEstPerDay } from '../../../../Types'

type CustomTooltipProps = {
  dataPoint: cloudEstPerDay
  grouping: string
}

const CustomTooltip = ({
  dataPoint,
  grouping,
}: CustomTooltipProps): ReactElement => {
  if (dataPoint?.x) {
    const dateLabel = {
      day: moment.utc(dataPoint.x).format('MMM DD, YYYY'),
      week: moment.utc(dataPoint.x).format('[Week] w, MMM'),
      month: moment.utc(dataPoint.x).format('MMM YYYY'),
      quarter: moment.utc(dataPoint.x).format('Qo [Quarter] YYYY'),
      year: moment.utc(dataPoint.x).format('YYYY'),
    }
    return (
      <div style={{ padding: '10px' }}>
        <div>
          <b>{dateLabel[grouping]}</b>
        </div>
        <div>
          {dataPoint.y} metric tons CO2e
          {dataPoint.usesAverageCPUConstant && '*'}
        </div>
        <div>{dataPoint.kilowattHours} kilowatt hrs</div>
        <div>${dataPoint.cost} cost</div>
      </div>
    )
  }
  return <div />
}

export default CustomTooltip
