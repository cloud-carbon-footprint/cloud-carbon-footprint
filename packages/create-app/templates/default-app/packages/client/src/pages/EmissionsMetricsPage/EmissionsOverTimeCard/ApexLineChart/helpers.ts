/*
 * Â© 2021 Thoughtworks, Inc.
 */

import { equals } from 'ramda'
import { EstimationResult } from '@cloud-carbon-footprint/common'
import { DateRange } from '../../../../Types'

const formatDateToTime = (timestamp: string | Date) =>
  timestamp instanceof Date
    ? timestamp.getTime()
    : new Date(timestamp).getTime()

const sortByDate = (data: EstimationResult[]): EstimationResult[] => {
  return data.sort((a: EstimationResult, b: EstimationResult) => {
    return formatDateToTime(a.timestamp) - formatDateToTime(b.timestamp)
  })
}

const filterBy = (
  data: EstimationResult[],
  range: DateRange,
  defaultRange: DateRange,
): EstimationResult[] => {
  if (!range.min || !range.max) return data
  if (equals(range, defaultRange)) return data

  const min = formatDateToTime(range.min)
  const max = formatDateToTime(range.max)
  return data.filter((a: EstimationResult) => {
    const result = formatDateToTime(a.timestamp)
    return result >= min && result <= max
  })
}

export { sortByDate, filterBy }
