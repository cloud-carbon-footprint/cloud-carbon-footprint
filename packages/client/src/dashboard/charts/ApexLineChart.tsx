/*
 * Â© 2021 ThoughtWorks, Inc.
 */

import React, { FunctionComponent, useEffect } from 'react'
import { renderToStaticMarkup } from 'react-dom/server'
import { useTheme } from '@material-ui/core/styles'
import { GetApp, PanTool, RotateLeft, ZoomIn } from '@material-ui/icons'
import { CustomTooltip } from './CustomTooltip'

import { getChartColors } from '../../themes'
import { sumServiceTotals, getMaxOfDataSeries } from '../transformData'
import { EstimationResult } from '@cloud-carbon-footprint/common'
import Chart from 'react-apexcharts'
import ApexCharts from 'apexcharts'
import * as _ from 'lodash'

export interface DateRange {
  min: Date | null
  max: Date | null
}

type LegendToggle = {
  [key: string]: boolean
}

const formatDateToTime = (timestamp: string | Date) =>
  timestamp instanceof Date
    ? timestamp.getTime()
    : new Date(timestamp).getTime()

export const sortByDate = (data: EstimationResult[]): EstimationResult[] => {
  return data.sort((a: EstimationResult, b: EstimationResult) => {
    return formatDateToTime(a.timestamp) - formatDateToTime(b.timestamp)
  })
}

export const filterBy = (
  data: EstimationResult[],
  range: DateRange,
  defaultRange: DateRange,
): EstimationResult[] => {
  if (!range.min || !range.max) return data
  if (_.isEqual(range, defaultRange)) return data

  const min = formatDateToTime(range.min)
  const max = formatDateToTime(range.max)
  return data.filter((a: EstimationResult) => {
    const result = formatDateToTime(a.timestamp)
    return result >= min && result <= max
  })
}

export const ApexLineChart: FunctionComponent<ApexLineChartProps> = ({
  data,
}) => {
  const theme = useTheme()
  const [dateRange, setDateRange] = React.useState<DateRange>({
    min: null,
    max: null,
  })
  const [chartData, setChartData] = React.useState<EstimationResult[]>([])
  const [defaultRange, setDefaultRange] = React.useState<DateRange>({
    min: null,
    max: null,
  })
  const [toggledSeries, setToggledSeries] = React.useState<LegendToggle[]>([
    { CO2e: true },
    { 'Kilowatt Hours': false },
    { Cost: false },
  ])

  useEffect(() => {
    const newSortedData = sortByDate(data)
    const newDefaultRange = {
      min: newSortedData[0]?.timestamp
        ? new Date(newSortedData[0]?.timestamp)
        : null,
      max: newSortedData[newSortedData.length - 1]?.timestamp
        ? new Date(newSortedData[newSortedData.length - 1]?.timestamp)
        : null,
    }

    if (!_.isEqual(chartData, newSortedData)) setChartData(newSortedData)
    if (
      newDefaultRange.min instanceof Date &&
      newDefaultRange.max instanceof Date
    ) {
      if (!_.isEqual(defaultRange, newDefaultRange)) {
        setDateRange(newDefaultRange)
        setDefaultRange(newDefaultRange)
      }
    }
  }, [data])

  const filteredByZoomRange = filterBy(chartData, dateRange, defaultRange)
  // We need to get the HTML string version of these icons since ApexCharts doesn't take in custom React components.
  // Why, you might ask? Don't ask me, ask ApexCharts.
  const GetAppIconHTML = renderToStaticMarkup(<GetApp />)
  const PanToolIconHTML = renderToStaticMarkup(<PanTool />)
  const RotateLeftIconHTML = renderToStaticMarkup(<RotateLeft />)
  const ZoomInIconHTML = renderToStaticMarkup(<ZoomIn />)

  const cloudEstimationData = sumServiceTotals(filteredByZoomRange)
  const co2SeriesData = cloudEstimationData.co2Series
  const kilowattHoursSeriesData = cloudEstimationData.kilowattHoursSeries
  const costSeriesData = cloudEstimationData.costSeries
  const maxCO2e = getMaxOfDataSeries(co2SeriesData)
  const maxKilowattHours = getMaxOfDataSeries(kilowattHoursSeriesData)
  const maxCost = getMaxOfDataSeries(costSeriesData)

  useEffect(() => {
    ApexCharts.exec('lineChart', 'updateSeries', [
      {
        name: 'CO2e',
        data: co2SeriesData,
      },
      {
        name: 'Kilowatt Hours',
        data: kilowattHoursSeriesData,
      },
      {
        name: 'Cost',
        data: costSeriesData,
      },
    ])

    toggledSeries.forEach((legendToggle: LegendToggle) => {
      const [seriesKey, toggleValue] = Object.entries(legendToggle)[0]
      toggleValue
        ? ApexCharts.exec('lineChart', 'showSeries', [seriesKey])
        : ApexCharts.exec('lineChart', 'hideSeries', [seriesKey])
    })
  }, [co2SeriesData, kilowattHoursSeriesData, costSeriesData, toggledSeries])

  const colors = getChartColors(theme)
  const [blue, yellow, green] = [colors[0], colors[5], colors[8]]

  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  const options: any = {
    chart: {
      events: {
        beforeZoom: (chart: unknown, { xaxis }: { xaxis: DateRange }) => {
          const newFilteredData = filterBy(data, xaxis, defaultRange)

          if (newFilteredData.length >= 2) {
            setDateRange(xaxis)
            return {
              xaxis,
            }
          }
          return {
            dateRange,
          }
        },
        beforeResetZoom: () => {
          setDateRange(defaultRange)
        },
        legendClick: (chart: unknown, seriesIndex: number) => {
          const [seriesKey, toggledValue] = Object.entries(
            toggledSeries[seriesIndex],
          )[0]
          const toggleCheck = toggledSeries.filter(
            (series) => !!Object.values(series)[0],
          )
          const newToggledSeries = [...toggledSeries]
          newToggledSeries[seriesIndex] = {
            [seriesKey]: !toggledValue,
          }
          setToggledSeries(newToggledSeries)
          if (
            toggleCheck.length === 1 &&
            Object.keys(toggleCheck[0])[0] === seriesKey
          )
            setToggledSeries(toggledSeries)
        },
      },
      id: 'lineChart',
      background: theme.palette.background.paper,
      toolbar: {
        tools: {
          download: GetAppIconHTML,
          zoom: ZoomInIconHTML,
          zoomin: false,
          zoomout: false,
          pan: PanToolIconHTML,
          reset: RotateLeftIconHTML,
        },
      },
    },
    colors: [blue, yellow, green],
    height: '500px',
    series: [
      {
        name: 'CO2e',
        data: co2SeriesData,
      },
      {
        name: 'Kilowatt Hours',
        data: kilowattHoursSeriesData,
      },
      {
        name: 'Cost',
        data: costSeriesData,
      },
    ],
    stroke: {
      width: 1,
    },
    theme: {
      mode: theme.palette.type,
    },
    tooltip: {
      shared: true,
      custom: function ({ dataPointIndex }: { dataPointIndex: number }) {
        return renderToStaticMarkup(
          <CustomTooltip
            data={co2SeriesData}
            dataPointIndex={dataPointIndex}
          />,
        )
      },
    },
    title: {
      text: 'Cloud Usage',
      offsetY: -8,
      style: {
        fontSize: '24px',
      },
    },
    xaxis: {
      type: 'datetime',
      title: {
        text: '',
        offsetY: 18,
        style: {
          fontSize: '15px',
        },
      },
    },
    yaxis: [
      {
        max: 1.1 * maxCO2e,
        title: {
          text: 'CO2e (metric tons)',
          offsetX: -8,
          style: {
            fontSize: '15px',
          },
        },
        tickAmount: 10,
        decimalsInFloat: 3,
      },
      {
        max: 1.1 * maxKilowattHours,
        title: {
          text: 'Kilowatt Hours (kWh)',
          opposite: -8,
          style: {
            fontSize: '15px',
            color: yellow,
          },
        },
        tickAmount: 10,
        decimalsInFloat: 2,
        opposite: true,
        axisBorder: {
          show: true,
          color: yellow,
        },
        axisTicks: {
          show: false,
        },
        showAlways: false,
      },
      {
        max: 1.1 * maxCost,
        title: {
          text: 'Cost ($)',
          offsetX: 6,
          style: {
            fontSize: '15px',
            color: green,
          },
        },
        tickAmount: 10,
        decimalsInFloat: 2,
        opposite: true,
        axisBorder: {
          show: true,
          color: green,
          offsetX: -5,
        },
        axisTicks: {
          show: false,
        },
        showAlways: false,
      },
    ],
    grid: {
      padding: {
        bottom: +15,
        right: +20,
      },
    },
  }
  return (
    <Chart
      aria-label="apex-line-chart"
      options={options}
      series={options.series}
      type="line"
      height={options.height}
    />
  )
}

type ApexLineChartProps = {
  data: EstimationResult[]
}
