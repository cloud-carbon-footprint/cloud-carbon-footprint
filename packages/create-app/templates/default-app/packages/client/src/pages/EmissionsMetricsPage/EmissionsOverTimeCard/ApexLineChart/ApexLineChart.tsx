/*
 * Â© 2021 Thoughtworks, Inc.
 */

import React, { FunctionComponent, useEffect } from 'react'
import { equals } from 'ramda'
import moment, { unitOfTime } from 'moment'
import { renderToStaticMarkup } from 'react-dom/server'
import ApexCharts from 'apexcharts'
import Chart from 'react-apexcharts'
import { useTheme } from '@material-ui/core/styles'
import { GetApp, PanTool, RotateLeft, ZoomIn } from '@material-ui/icons'
import { EstimationResult } from '@cloud-carbon-footprint/common'
import { ApexChartProps, DateRange } from '../../../../Types'
import { getChartColors, CCFTheme } from '../../../../utils/themes'
import { sumServiceTotals, getMaxOfDataSeries } from '../../../../utils/helpers'
import { filterBy, sortByDate } from './helpers'
import CustomTooltip from './CustomTooltip'

type LegendToggle = {
  [key: string]: boolean
}

const ApexLineChart: FunctionComponent<ApexChartProps> = ({ data }) => {
  const theme = useTheme() as CCFTheme
  const colors = getChartColors(theme)
  const [blue, yellow, green] = [colors[0], colors[5], colors[8]]

  const [dateRange, setDateRange] = React.useState<DateRange>({
    min: null,
    max: null,
  })
  const [chartData, setChartData] = React.useState<EstimationResult[]>([])
  const [defaultRange, setDefaultRange] = React.useState<DateRange>({
    min: null,
    max: null,
  })
  const [toggledSeries, setToggledSeries] = React.useState<LegendToggle[]>([
    { CO2e: true },
    { 'Kilowatt Hours': false },
    { Cost: false },
  ])

  const filteredByZoomRange = filterBy(chartData, dateRange, defaultRange)

  // We need to get the HTML string version of these icons since ApexCharts doesn't take in custom React components.
  // Why, you might ask? Don't ask me, ask ApexCharts.
  const GetAppIconHTML = renderToStaticMarkup(<GetApp />)
  const PanToolIconHTML = renderToStaticMarkup(<PanTool />)
  const RotateLeftIconHTML = renderToStaticMarkup(<RotateLeft />)
  const ZoomInIconHTML = renderToStaticMarkup(<ZoomIn />)

  const cloudEstimationData = sumServiceTotals(filteredByZoomRange)
  const co2SeriesData = cloudEstimationData.co2Series
  const kilowattHoursSeriesData = cloudEstimationData.kilowattHoursSeries
  const costSeriesData = cloudEstimationData.costSeries
  const maxCO2e = getMaxOfDataSeries(co2SeriesData)
  const maxKilowattHours = getMaxOfDataSeries(kilowattHoursSeriesData)
  const maxCost = getMaxOfDataSeries(costSeriesData)

  const grouping = data[0]?.groupBy || 'day'
  const dateFormat = {
    day: 'MMM DD, YYYY',
    week: '[Week] w, MMM',
    month: 'MMM YYYY',
    quarter: 'Qo [Quarter] YYYY',
    year: 'YYYY',
  }

  useEffect(() => {
    const newSortedData = sortByDate(data)
    const min = newSortedData[0]?.timestamp
      ? new Date(newSortedData[0]?.timestamp)
      : null
    const endDate = new Date(newSortedData[newSortedData.length - 1]?.timestamp)
    const max = endDate
      ? moment(endDate)
          .add(1, `${grouping}s` as unitOfTime.DurationConstructor)
          .toDate()
      : null
    const newDefaultRange = {
      min,
      max,
    }

    if (!equals(chartData, newSortedData)) setChartData(newSortedData)
    if (
      newDefaultRange.min instanceof Date &&
      newDefaultRange.max instanceof Date
    ) {
      if (!equals(defaultRange, newDefaultRange)) {
        setDateRange(newDefaultRange)
        setDefaultRange(newDefaultRange)
      }
    }
  }, [data])

  useEffect(() => {
    ApexCharts.exec('lineChart', 'updateSeries', [
      {
        name: 'CO2e',
        data: co2SeriesData,
      },
      {
        name: 'Kilowatt Hours',
        data: kilowattHoursSeriesData,
      },
      {
        name: 'Cost',
        data: costSeriesData,
      },
    ])

    ApexCharts.exec('lineChart', 'updateOptions', [
      {
        xaxis: {
          category: global.labels,
        },
      },
    ])

    toggledSeries.forEach((legendToggle: LegendToggle) => {
      const [seriesKey, toggleValue] = Object.entries(legendToggle)[0]
      toggleValue
        ? ApexCharts.exec('lineChart', 'showSeries', [seriesKey])
        : ApexCharts.exec('lineChart', 'hideSeries', [seriesKey])
    })
  }, [co2SeriesData, kilowattHoursSeriesData, costSeriesData, toggledSeries])

  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  const options: any = {
    markers: {
      size: 5,
    },
    chart: {
      events: {
        beforeZoom: (chart: unknown, { xaxis }: { xaxis: DateRange }) => {
          const newFilteredData = filterBy(data, xaxis, defaultRange)

          if (newFilteredData.length >= 2) {
            setDateRange(xaxis)
            return {
              xaxis,
            }
          }
          return {
            dateRange,
          }
        },
        beforeResetZoom: () => {
          setDateRange(defaultRange)
        },
        legendClick: (chart: unknown, seriesIndex: number) => {
          const [seriesKey, toggledValue] = Object.entries(
            toggledSeries[seriesIndex],
          )[0]
          const toggleCheck = toggledSeries.filter(
            (series) => !!Object.values(series)[0],
          )
          const newToggledSeries = [...toggledSeries]
          newToggledSeries[seriesIndex] = {
            [seriesKey]: !toggledValue,
          }
          setToggledSeries(newToggledSeries)
          if (
            toggleCheck.length === 1 &&
            Object.keys(toggleCheck[0])[0] === seriesKey
          )
            setToggledSeries(toggledSeries)
        },
      },
      id: 'lineChart',
      background: theme.palette.background.paper,
      toolbar: {
        tools: {
          download: GetAppIconHTML,
          zoom: ZoomInIconHTML,
          zoomin: false,
          zoomout: false,
          pan: PanToolIconHTML,
          reset: RotateLeftIconHTML,
        },
      },
    },
    colors: [blue, yellow, green],
    height: '500px',
    series: [
      {
        name: 'CO2e',
        data: co2SeriesData,
      },
      {
        name: 'Kilowatt Hours',
        data: kilowattHoursSeriesData,
      },
      {
        name: 'Cost',
        data: costSeriesData,
      },
    ],
    stroke: {
      width: 1,
    },
    theme: {
      mode: theme.palette.type,
    },
    tooltip: {
      shared: true,
      custom: function ({ dataPointIndex }: { dataPointIndex: number }) {
        return renderToStaticMarkup(
          <CustomTooltip
            dataPoint={co2SeriesData[dataPointIndex]}
            grouping={grouping}
          />,
        )
      },
    },
    title: {
      text: 'Cloud Usage',
      offsetY: -8,
      style: {
        fontSize: '24px',
      },
    },
    xaxis: {
      type: 'category',
      tickAmount: 'dataPoints',
      title: {
        text: '',
        offsetY: 18,
        style: {
          fontSize: '15px',
        },
      },
      labels: {
        formatter: function (val) {
          return moment(val).add(1, `d`).format(dateFormat[grouping])
        },
        style: {
          display: 'contents !important',
        },
      },
    },
    yaxis: [
      {
        max: 1.1 * maxCO2e,
        title: {
          text: 'CO2e (metric tons)',
          offsetX: -8,
          style: {
            fontSize: '15px',
          },
        },
        tickAmount: 10,
        decimalsInFloat: 3,
      },
      {
        max: 1.1 * maxKilowattHours,
        title: {
          text: 'Kilowatt Hours (kWh)',
          opposite: -8,
          style: {
            fontSize: '15px',
            color: yellow,
          },
        },
        tickAmount: 10,
        decimalsInFloat: 2,
        opposite: true,
        axisBorder: {
          show: true,
          color: yellow,
        },
        axisTicks: {
          show: false,
        },
        showAlways: false,
      },
      {
        max: 1.1 * maxCost,
        title: {
          text: 'Cost ($)',
          offsetX: 6,
          style: {
            fontSize: '15px',
            color: green,
          },
        },
        tickAmount: 10,
        decimalsInFloat: 2,
        opposite: true,
        axisBorder: {
          show: true,
          color: green,
          offsetX: -5,
        },
        axisTicks: {
          show: false,
        },
        showAlways: false,
      },
    ],
    grid: {
      padding: {
        bottom: +15,
        right: +20,
      },
    },
  }

  return (
    <Chart
      aria-label="apex-line-chart"
      options={options}
      series={options.series}
      type="line"
      height={options.height}
    />
  )
}

export default ApexLineChart
