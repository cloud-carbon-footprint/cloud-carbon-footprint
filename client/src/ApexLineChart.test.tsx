import React from 'react'
import { create } from 'react-test-renderer'
import moment from 'moment'

import { ApexLineChart } from './ApexLineChart'

jest.mock('./themes')

describe('ApexLineChart', () => {
  it('renders with correct configuration', () => {
    const data = [
      {
        timestamp: moment('2019-08-10T00:00:00.000Z').toDate(),
        serviceEstimates: [
          {
            timestamp: moment('2019-08-10T00:00:00.000Z').toDate(),
            serviceName: 'ebs',
            wattHours: 0,
            co2e: 5,
            cost: 0,
            region: 'us-west-2',
          },
          {
            timestamp: moment('2019-08-10T00:00:00.000Z').toDate(),
            serviceName: 's3',
            wattHours: 0,
            co2e: 0,
            cost: 0,
            region: 'us-west-2',
          },
          {
            timestamp: moment('2019-08-10T00:00:00.000Z').toDate(),
            serviceName: 'ec2',
            wattHours: 0,
            co2e: 2,
            cost: 0,
            region: 'us-west-2',
          },
        ],
      },
    ]
    const root = create(<ApexLineChart data={data} />)

    expect(root.toJSON()).toMatchSnapshot()
  })
})
