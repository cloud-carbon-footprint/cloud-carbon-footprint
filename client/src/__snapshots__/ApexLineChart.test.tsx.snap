// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ApexLineChart renders with correct configuration 1`] = `
<div
  height="500px"
  options={
    Object {
      "chart": Object {
        "id": "apexchart-example",
        "toolbar": Object {
          "tools": Object {
            "customIcons": Array [],
            "zoomin": false,
            "zoomout": false,
          },
        },
      },
      "height": "500px",
      "markers": Object {
        "size": 5,
      },
      "series": Array [
        Object {
          "data": Array [
            Object {
              "x": 2019-08-10T00:00:00.000Z,
              "y": 7,
            },
          ],
          "name": "co2e",
        },
      ],
      "stroke": Object {
        "width": 2,
      },
      "title": Object {
        "offsetY": -8,
        "style": Object {
          "fontSize": "24px",
        },
        "text": "Cloud Emissions",
      },
      "xaxis": Object {
        "convertedCatToNumeric": false,
        "labels": Object {
          "formatter": [Function],
        },
        "title": Object {
          "offsetY": 8,
          "style": Object {
            "fontSize": "15px",
          },
          "text": "",
        },
        "type": "datetime",
      },
      "yaxis": Object {
        "title": Object {
          "offsetX": -8,
          "style": Object {
            "fontSize": "15px",
          },
          "text": "CO2e (kg)",
        },
      },
    }
  }
  series={
    Array [
      Object {
        "data": Array [
          Object {
            "x": 2019-08-10T00:00:00.000Z,
            "y": 7,
          },
        ],
        "name": "co2e",
      },
    ]
  }
  type="line"
  width="100%"
/>
`;
